satellite/repair/repairer/ec.go

```diff
@@ -0,0 +1,449 @@
+	}
+
+	if !unique(limits) {
+		return nil, nil, Error.New("duplicated nodes are not allowed")
+	}
+
+	readers, err := eestream.EncodeReader(ctx, ec.log, ioutil.NopCloser(data), rs)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	type info struct {
+		i    int
+		err  error
+		hash *pb.PieceHash
+	}
+	infos := make(chan info, pieceCount)
+
+	psCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for i, addressedLimit := range limits {
+		go func(i int, addressedLimit *pb.AddressedOrderLimit) {
+			hash, err := ec.putPiece(psCtx, ctx, addressedLimit, privateKey, readers[i], expiration)
+			infos <- info{i: i, err: err, hash: hash}
+		}(i, addressedLimit)
+	}
+	ec.log.Info("Starting a timer for repair so that the number of pieces will be closer to the success threshold",
+		zap.Duration("Timer", timeout),
+		zap.String("Path", path),
+		zap.Int("Node Count", nonNilCount(limits)),
+		zap.Int("Optimal Threshold", rs.OptimalThreshold()),
+	)
+
+	var successfulCount, failureCount, cancellationCount int32
+	timer := time.AfterFunc(timeout, func() {
+		if ctx.Err() != context.Canceled {
+			ec.log.Info("Timer expired. Canceling the long tail...",
+				zap.String("Path", path),
+				zap.Int32("Successfully repaired", atomic.LoadInt32(&successfulCount)),
+			)
+			cancel()
+		}
+	})
+
+	successfulNodes = make([]*pb.Node, pieceCount)
+	successfulHashes = make([]*pb.PieceHash, pieceCount)
+
+	for range limits {
+		info := <-infos
+
+		if limits[info.i] == nil {
+			continue
+		}
+
+		if info.err != nil {
+			if !errs2.IsCanceled(info.err) {
+				failureCount++
+			} else {
+				cancellationCount++
+			}
+			ec.log.Debug("Repair to storage node failed",
+				zap.String("Path", path),
+				zap.String("NodeID", limits[info.i].GetLimit().StorageNodeId.String()),
+				zap.Error(info.err),
+			)
+			continue
+		}
+
+		successfulNodes[info.i] = &pb.Node{
+			Id:      limits[info.i].GetLimit().StorageNodeId,
+			Address: limits[info.i].GetStorageNodeAddress(),
+		}
+		successfulHashes[info.i] = info.hash
+		successfulCount++
+	}
+
+	// Ensure timer is stopped
+	_ = timer.Stop()
+
+	// TODO: clean up the partially uploaded segment's pieces
+	defer func() {
+		select {
+		case <-ctx.Done():
+			err = Error.New("repair cancelled")
+		default:
+		}
+	}()
+
+	if successfulCount == 0 {
+		return nil, nil, Error.New("repair %v to all nodes failed", path)
+	}
+
+	ec.log.Info("Successfully repaired",
+		zap.String("Path", path),
+		zap.Int32("Success Count", atomic.LoadInt32(&successfulCount)),
+	)
+
+	mon.IntVal("repair_segment_pieces_total").Observe(int64(pieceCount))
+	mon.IntVal("repair_segment_pieces_successful").Observe(int64(successfulCount))
+	mon.IntVal("repair_segment_pieces_failed").Observe(int64(failureCount))
+	mon.IntVal("repair_segment_pieces_canceled").Observe(int64(cancellationCount))
+
+	return successfulNodes, successfulHashes, nil
+}
+
+// copied from ecclient
```

Comment:
are we keeping these comments long-term? I'm not sure if they're necessary

Reply:
I think they can be removed. I was hoping we could get rid of duplicate code here and in ecclient, but I don't know of a good way.

Reply:
Comments removed

Reply:
maybe someone else has thoughts on this? maybe could make a ec_helpers.go file or something?